index.bs

```diff
@@ -7,14 +7,110 @@ Group: browser-testing-tools
+remote control of user agents. This specification extends WebDriver by
+introducing bidirectional communication. In place of the strict
+command/response format of WebDriver, this permits events to stream
+from the user agent to the controlling software, better matching the
+evented nature of the browser DOM.
+
+Transport {#transport}
+======================
+
+Message transport is provided using a [[RFC6455|WebSocket]]
+connection.
+
+A [=remote end=] has a <dfn>connection</dfn>, which is a network connection, and
+a <dfn>websocket url</dfn>, which is a string. These are both initially null.
+
+When data is recieved on [=connection=] it must be processed according to the
+requirements of the [[!RFC6455|WebSockets]] specification. When [=a WebSocket
+message has been received=], an implementation must [=handle an incoming
+message=]. When [=the WebSocket closing handshake is started=] an implementation
+must [=end the session=].
+
+When required to <dfn>ensure a connection is established</dfn>, run
+the following steps:
+
+1.  If the [=websocket url=] is not null, return the
+    [=websocket url=]
+
+2.  Set up a network connection that listens on an implementation-defined
+    hostname <var>host</var> and port <var>port</var>. Set the [=connection=] to
+    this network connection.
+
+3.  Let <var>WebSocket URL</var> be the result of constructing a [=WebSocket
+    ws-URI=] with host <var>host</var>, port <var>port</var>, and empty path an
+    query.
+
+4.  Set the [=websocket url=] to <var>WebSocket URL</var>.
+
+5.  Return <var>WebSocket URL</var>.
+
+
+Note: Typically the hostname in the above steps will be
+"<code>localhost</code>".
+
+When required to <dfn>handle an incoming message</dfn>, run the following steps:
+
+Issue: Figure this out.
+
+When required to <dfn>end the session</dfn>, run the following steps:
+
+1. Close the underlying network connection [=connection=].
+
+2. Set [=connection=] and [=WebSocket URL=] to null.
+
+Issue: This should also reset any internal state
+
+Note: This does not end any [=session=].
+
+Establishing a Connection {#estabishing}
+========================================
+
+WebDriver clients opt in to a bidirectional connection by requesting a
+capability with the name "<code>bidi</code>" and value
+<code>true</code>.
+
+This specification defines an
+[=additional webdriver capability=] with the [=capability name=] "<code>bidi</code>".
+
+The [=additional capability deserialization algorithm=] for the
+"<code>bidi</code> capability, with argument <var>value</var> is:
+ 1. If <var>value</var> is not a boolean, return [=error=] with [=error
+    code|code=] [=invalid argument=].
+ 2. Return [=success=] with data <var>value</var>.
+
+The [=matched capability serialization algorithm=] for the `bidi` capability,
+with argument <var>value</var> is:
+ 1. If value is <code>false</code>, return [=success=] with data
+    <code>null</code>.
+ 2. Let <var>websocket url</var> be the result of running the steps to
```

Comment:
Seeing how this is called from https://w3c.github.io/webdriver/#dfn-matching-capabilities, the naming of the this concept, matched capability serialization algorithm, seems pretty odd. Would "additional capability matching algorithm" be an improvement?

Reply:
I don't think I follow which thing you want to be renamed.

Reply:
My comment didn't end up where I intended. I mean "matched capability serialization algorithm" sounds odd. But maybe with the new extension point you're adding this will go away.